app/models/services/locators/base_locator.rb

```diff
@@ -8,4 +8,50 @@ def initialize locator_id
   def find_paper
     Paper.find_by self.column => self.locator_id
   end
+
+  def import_data_to_paper(paper,imported_data,source)
+    source = source.downcase
+
+    if source == "pubmed"
+      if paper.authors.empty?
+        names = imported_data['authors'].map {|a| a['name']}
+        existing_authors = Author.where(name: names)
+        existing_names = existing_authors.map(&:name)
+        new_authors = (names - existing_names).map {|a| Author.create name: a}
+      end
+
+      doi = imported_data['elocationid']
+      if doi.blank?
+        doi = imported_data['articleids'].find {|id| id['idtype'] == 'doi' }['value']
+      end
+
+      doi = doi.sub(/^doi: /, "")
+
+      paper.pubmed_id ||= imported_data['uid']
+      paper.title ||= imported_data['title']
+      paper.published_at ||=  imported_data['pubdate']
+      paper.authors ||= (existing_authors + new_authors)
+      paper.abstract ||= Pubmed.new.get_abstract(imported_data['uid'])
+      paper.doi ||= doi
+      paper.publication ||= imported_data['source']
+    end
+
+    if source == "crossref"
+      if paper.authors.empty? && imported_data['author']
+        names = imported_data['author'].map {|a| "#{a['given']} #{a['family']}"}
+        existing_authors = Author.where(name: names)
+        existing_names = existing_authors.map(&:name)
+        new_authors = (names - existing_names).map {|a| Author.create name: a}
+        paper.authors = (existing_authors + new_authors)
+      end
+
+      paper.title ||= imported_data['title'].first
+      paper.published_at ||=  imported_data['created']['date-time'].to_date
+      paper.doi ||= imported_data['DOI']
+      paper.publication ||= imported_data['publisher']
+      paper.link ||= imported_data['link'].first['URL'] if imported_data['link']
+    end
+
+    return paper
+  end
```

Comment:
I'm not comfortable with this solution. Besides which, it almost certainly should not be in the base class. We definitely need to create different mappings per service, but those mappings should not be here.